sys/xtimer/xtimer_core.c

```diff
@@ -265,8 +162,9 @@ static void _add_timer_to_list(xtimer_t **list_head, xtimer_t *timer)
 static void _add_timer_to_long_list(xtimer_t **list_head, xtimer_t *timer)
 {
     while (*list_head
-           && (((*list_head)->long_target < timer->long_target)
-           || (((*list_head)->long_target == timer->long_target) && ((*list_head)->target <= timer->target)))) {
+        && (((*list_head)->long_offset < timer->long_offset)
+        || (((*list_head)->long_offset == timer->long_offset)
+            && ((*list_head)->start_time + (*list_head)->offset - timer->start_time <= timer->offset)))) {
```

Comment:
Could you change it to:
`&& ((*list_head)->start_time + (*list_head)->offset <= timer->start_time + timer->offset )))) {`

Reply:
No. We should compare two offsets, not two absolute time. When Offset1 < Offset2, timer 1 expires early than timer 2. But when AbsoluteTime1 < AbsoluteTime2, we don't know if timer 1 expires earlier or not.

Reply:
Is not this: `&& ((*list_head)->start_time + (*list_head)->offset - timer->start_time <= timer->offset)))) {`
the same as:
`&& ((*list_head)->start_time + (*list_head)->offset <= timer->start_time + timer->offset )))) {`

it just looks better.

Reply:
"start_time1 + offset1" is an absolute target time and comparing two absolute target time will cause severe errors. Instread, "start_time1 + offset1 - start_time2" is a remaining offset, which is safe to compare.

Reply:
For example, when start_time1 = 0xFFFFFFF0, offset1 = 5, start_time2 = 0xFFFFFFF0, and offset2 = 0x0F, timer1 must be expired earlier than timer2. However, given that start_time1 + offset1 = 0xFFFFFFF5 and start_time2 + offset2 = 0, timer2's target time seems smaller, which causes an error.

Reply:
In contrast, "start_time1 + offset1 - start_time2" = 0x05 and offset2 = 0x0F. Comparing these two offsets gives the right result.

Reply:
Ok I understand. I did not think of that. Thank you.

Maybe it might be better to introduce a new function to compare two timers. It makes the code better to read.

Reply:
Good point. Done.